lib/charms/opensearch/v0/opensearch_nodes_exclusions.py

```diff
@@ -70,55 +72,134 @@ def delete_current(self) -> None:
-            return True
+            logger.debug(f"Added voting for {to_add}: SUCCESS")
+            return to_add is not None
         except OpenSearchHttpError:
+            logger.debug(f"Added voting for {to_add}: FAILED")
             return False
 
-    def _delete_voting(self) -> bool:
+    def _delete_voting(self, to_remove: Optional[Set[str]] = None) -> Optional[Set[str]]:
         """Remove all the voting exclusions - cannot target 1 exclusion at a time."""
         # "wait_for_removal" is VERY important, it removes all voting configs immediately
         # and allows any node to return to the voting config in the future
+        original_exclusions = exclusions = self._fetch_voting_exclusions()
         try:
             self._opensearch.request(
                 "DELETE",
                 "/_cluster/voting_config_exclusions?wait_for_removal=false",
                 alt_hosts=self._charm.alt_hosts,
                 resp_status_code=True,
             )
-            return True
+            logger.debug("Removed voting")
+            remove_set = to_remove or set([self._node.name])
+            nodes = set(remove_set)
+            for node in nodes:
+                if node in exclusions:
+                    exclusions.remove(node)
+                    remove_set.remove(node)
+            if exclusions and exclusions != original_exclusions:
+                logger.debug("Removed voting, adding back voting")
+                self._add_voting(exclusions)
+            return remove_set is None
         except OpenSearchHttpError:
+            logger.debug("Removed voting: FAILED")
             return False
 
-    def _add_allocations(
+    def _fetch_voting_exclusions(self) -> Set[str]:
+        """Fetch the registered voting exclusions."""
+        try:
+            resp = self._opensearch.request(
+                "GET",
+                "/_cluster/state/metadata/voting_config_exclusions",
+                alt_hosts=self._charm.alt_hosts,
+            )
+            return set(
+                sorted(
+                    [
+                        node["node_name"]
+                        for node in resp["metadata"]["cluster_coordination"][
+                            "voting_config_exclusions"
+                        ]
+                    ]
+                )
+            )
+        except (OpenSearchHttpError, KeyError) as e:
+            logger.warning(f"Failed to fetch voting exclusions: {e}")
+            # no voting exclusion set
+            return {}
+
+    def add_allocations(
         self, allocations: Optional[Set[str]] = None, override: bool = False
     ) -> bool:
         """Register new allocation exclusions."""
         try:
             existing = set() if override else self._fetch_allocations()
             all_allocs = existing.union(
-                allocations if allocations is not None else {self._node.name}
+                allocations if allocations is not None else set([self._node.name])
```

Comment:
I believe this is a more expensive call, building a list, building a set, as opposed to what we had